# Capítulo 24 — Primeras preguntas de análisis

## De preparar datos a analizarlos

En los capítulos anteriores trabajamos sobre la preparación de datos: revisamos estructuras, detectamos problemas, transformamos columnas, tratamos valores faltantes, convertimos tipos de datos y validamos el resultado final.

A partir de ahora el foco cambia.

Ya no vamos a mirar un `DataFrame` principalmente para corregirlo, sino para empezar a hacer preguntas sobre los datos.

Analizar datos no significa aplicar muchos comandos de manera automática. Significa formular preguntas, buscar respuestas en el dataset, interpretar los resultados y decidir qué nuevas preguntas aparecen a partir de esas respuestas.

En este capítulo vamos a comenzar con preguntas simples:

* ¿Cuántos registros tenemos?
* ¿Qué variables contiene el dataset?
* ¿Qué valores son más frecuentes?
* ¿Cuáles son los valores mínimos, máximos o promedio?
* ¿Qué primeras ideas podemos obtener antes de hacer análisis más complejos?

Todavía no vamos a construir un análisis completo. El objetivo es dar el primer paso desde un dataset disponible hacia preguntas concretas de análisis.


## Usar un dataset de práctica

Para comenzar esta nueva etapa vamos a usar un dataset de práctica incluido en una librería de Python.

Muchas librerías orientadas al análisis de datos, la visualización o el aprendizaje automático incluyen datasets pequeños y conocidos. Estos datasets se usan para aprender, probar herramientas, comparar resultados y practicar sin depender siempre de archivos externos.

Esto tiene una ventaja importante: podemos concentrarnos en las preguntas de análisis sin comenzar otra vez por la descarga, la carga manual o la limpieza profunda de los datos.

En este caso vamos a usar el dataset `tips`, disponible desde la librería Seaborn.

Este dataset contiene información sobre cuentas y propinas en un restaurante. Cada fila representa una mesa atendida, e incluye datos como el total de la cuenta, la propina, el día, el horario y la cantidad de personas.

Aunque es un dataset pequeño, resulta muy útil para practicar análisis exploratorio porque combina variables numéricas y categóricas.


In [29]:
import pandas as pd
import seaborn as sns

df = sns.load_dataset("tips")

df.head()

,total_bill,tip,sex,smoker,day,time,size
0,16.99,1.01,Female,No,Sun,Dinner,2
1,10.34,1.66,Male,No,Sun,Dinner,3
2,21.01,3.50,Male,No,Sun,Dinner,3
3,23.68,3.31,Male,No,Sun,Dinner,2
4,24.59,3.61,Female,No,Sun,Dinner,4


La tabla muestra las primeras filas del dataset.

Cada registro corresponde a una cuenta de restaurante. Algunas columnas contienen valores numéricos, como el total de la cuenta, la propina o la cantidad de personas. Otras columnas contienen categorías, como el día, el horario o si la mesa correspondía a una persona fumadora o no fumadora.

En esta etapa no vamos a limpiar el dataset. Lo vamos a usar como punto de partida para formular y responder primeras preguntas de análisis.


## Reconocer el tamaño del dataset

Antes de formular preguntas de análisis, conviene saber con qué volumen de datos estamos trabajando.

Vamos a revisar cuántas filas y columnas tiene el dataset.

In [30]:
df.shape

(244, 7)

El dataset contiene **244 filas y 7 columnas**.

Esto confirma que estamos trabajando con un conjunto de datos pequeño, adecuado para practicar análisis exploratorio sin que el volumen de información complique la lectura de los resultados.

Ahora vamos a revisar qué columnas están disponibles. No se trata solo de mirar nombres: cada columna representa una posible variable para analizar.

In [31]:
df.columns

Index(['total_bill', 'tip', 'sex', 'smoker', 'day', 'time', 'size'], dtype='object')

Los nombres de las columnas nos muestran qué información está disponible para analizar.

Las columnas disponibles son:

- `total_bill`: total de la cuenta.
- `tip`: propina.
- `sex`: sexo de la persona registrada como pagadora o responsable de la cuenta.
- `smoker`: si la mesa correspondía a personas fumadoras o no fumadoras.
- `day`: día de la semana.
- `time`: momento del día.
- `size`: cantidad de personas en la mesa.

A partir de estas variables ya podemos formular primeras preguntas de análisis: cuánto se consumió, cuánto se dejó de propina, qué días aparecen en el dataset, qué horarios se registraron y qué tamaño tenían las mesas.

También conviene revisar cómo interpreta Pandas cada columna:

In [32]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 244 entries, 0 to 243
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype   
---  ------      --------------  -----   
 0   total_bill  244 non-null    float64 
 1   tip         244 non-null    float64 
 2   sex         244 non-null    category
 3   smoker      244 non-null    category
 4   day         244 non-null    category
 5   time        244 non-null    category
 6   size        244 non-null    int64   
dtypes: category(4), float64(2), int64(1)
memory usage: 7.4 KB


La salida de `info()` confirma que el dataset tiene **244 registros** y que no aparecen valores faltantes en ninguna columna: todas tienen **244 valores no nulos**.

También vemos que Pandas interpreta las columnas de esta manera:

- `total_bill` y `tip` son columnas numéricas con decimales.
- `size` es una columna numérica entera.
- `sex`, `smoker`, `day` y `time` son columnas categóricas.

Esto es importante para el análisis porque no todas las preguntas se responden del mismo modo. Con las columnas numéricas podremos calcular totales, promedios, mínimos y máximos. Con las columnas categóricas podremos contar frecuencias y comparar grupos.

En otras palabras, el dataset ya nos ofrece dos caminos iniciales de análisis: resumir valores numéricos y contar categorías.

## Formular primeras preguntas

Antes de calcular indicadores, conviene transformar la curiosidad general en preguntas concretas.

En este dataset podríamos empezar con preguntas como:

- ¿Cuál fue el total promedio de las cuentas?
- ¿Cuál fue la propina promedio?
- ¿Cuál fue la cuenta más alta?
- ¿Qué días aparecen registrados?
- ¿Qué día aparece con más frecuencia?
- ¿Cuál es el tamaño de mesa más habitual?

Estas preguntas todavía son simples, pero marcan un cambio importante: ya no estamos revisando el dataset para prepararlo, sino para entender qué información contiene.

In [33]:
df[["total_bill", "tip", "size"]].describe()

,total_bill,tip,size
count,244.000000,244.000000,244.000000
mean,19.785943,2.998279,2.569672
std,8.902412,1.383638,0.951100
min,3.070000,1.000000,1.000000
25%,13.347500,2.000000,2.000000
50%,17.795000,2.900000,2.000000
75%,24.127500,3.562500,3.000000
max,50.810000,10.000000,6.000000


La tabla resume las tres columnas numéricas del dataset.

Podemos leer algunos primeros resultados:

- Hay **244 valores** en cada columna, lo que confirma que estas variables no tienen datos faltantes.
- El total promedio de las cuentas (`total_bill`) es de aproximadamente **19.79**.
- La propina promedio (`tip`) es de aproximadamente **3.00**.
- El tamaño promedio de las mesas (`size`) es de aproximadamente **2.57 personas**.

También podemos observar los valores extremos:

- La cuenta más baja fue de **3.07**.
- La cuenta más alta fue de **50.81**.
- La propina más baja fue de **1.00**.
- La propina más alta fue de **10.00**.
- El tamaño de las mesas va de **1** a **6** personas.

Esta salida ya nos permite obtener una primera imagen general del dataset. Sin embargo, todavía son resultados globales: sabemos cómo se comportan las columnas numéricas en conjunto, pero no sabemos si esos valores cambian según el día, el horario o el tipo de mesa.

## Contar valores en una columna categórica

Hasta ahora observamos resúmenes generales de las columnas numéricas. Ahora vamos a mirar una columna categórica.

Una primera pregunta posible es:

**¿Qué días aparecen registrados en el dataset y cuántas cuentas hay en cada uno?**

Para responderla podemos contar cuántas veces aparece cada valor en la columna `day`.

In [34]:
df["day"].value_counts()

,count
day,
Sat,87
Sun,76
Thur,62
Fri,19


La columna `day` muestra registros para cuatro días: `Thur`, `Fri`, `Sat` y `Sun`.

El día con más cuentas registradas es **sábado**, con **87 cuentas**. Luego aparecen **domingo**, con **76**, y **jueves**, con **62**. El día con menos registros es **viernes**, con **19 cuentas**.

Este resultado no significa necesariamente que el sábado sea “mejor” o “más rentable”. Por ahora solo sabemos que en el dataset hay más cuentas registradas ese día. Para hablar de ingresos, promedios o comportamiento de consumo necesitaríamos hacer otros cálculos.

## Contar registros por horario

También podemos preguntarnos en qué momento del día se registraron más cuentas.

La columna `time` indica si cada registro corresponde al almuerzo o a la cena. Contar sus valores nos permite ver cómo se distribuyen las cuentas entre esos dos momentos.

In [35]:
df["time"].value_counts()

,count
time,
Dinner,176
Lunch,68


La mayor parte de las cuentas registradas corresponde a **Dinner**, con **176 registros**. En cambio, **Lunch** aparece con **68 registros**.

Esto nos permite ver que el dataset tiene más observaciones de cenas que de almuerzos.

Esta salida solo describe la cantidad de registros por horario. Para analizar consumo, propinas o tamaño de mesa necesitaremos combinar esta información con columnas numéricas como `total_bill`, `tip` o `size`.

## Analizar el tamaño de las mesas

Otra pregunta inicial es cuántas personas había en cada mesa.

La columna `size` indica el tamaño de la mesa. Aunque es una columna numérica, en este caso también puede analizarse como una categoría: mesa de 1 persona, de 2 personas, de 3 personas, y así sucesivamente.

Contar estos valores nos permite saber qué tamaños de mesa aparecen con más frecuencia en el dataset.

In [36]:
df["size"].value_counts()

,count
size,
2,156
3,38
4,37
5,5
1,4
6,4


El tamaño de mesa más frecuente es **2 personas**, con **156 registros**.

Luego aparecen las mesas de **3 personas** y **4 personas**, con cantidades bastante similares: **38** y **37 registros** respectivamente.

Los tamaños menos frecuentes son las mesas de **1**, **5** y **6** personas.

Este resultado nos muestra que el dataset está dominado por mesas pequeñas, especialmente de dos personas. Esta información puede ser importante más adelante, porque el tamaño de la mesa puede influir en el total de la cuenta y también en la propina.

## Calcular indicadores simples

Además de contar registros, podemos calcular algunos indicadores numéricos básicos.

Por ejemplo, podemos preguntarnos:

- ¿Cuál fue el total facturado en todas las cuentas?
- ¿Cuánto se dejó de propina en total?
- ¿Cuál fue el promedio del total de cuenta?
- ¿Cuál fue el promedio de propina?

Estas preguntas todavía no comparan grupos ni categorías. Son indicadores generales del dataset completo.

In [37]:
total_facturado = df["total_bill"].sum()
total_propinas = df["tip"].sum()
promedio_cuenta = df["total_bill"].mean()
promedio_propina = df["tip"].mean()

total_facturado, total_propinas, promedio_cuenta, promedio_propina

(np.float64(4827.77),
 np.float64(731.5799999999999),
 np.float64(19.78594262295082),
 np.float64(2.99827868852459))

Los resultados obtenidos corresponden a cuatro indicadores generales del dataset:

- El total facturado es **4827.77**.
- El total de propinas es **731.58**.
- El promedio de cuenta es aproximadamente **19.79**.
- El promedio de propina es aproximadamente **3.00**.

Estos valores resumen el conjunto completo de datos, sin distinguir todavía por día, horario, tamaño de mesa u otras categorías.

La salida es correcta, pero no es muy cómoda para leer: aparece como una tupla de valores y Pandas/NumPy muestra algunos números con demasiados decimales. Por eso conviene construir una tabla pequeña de indicadores.

## Presentar indicadores en una tabla

Cuando calculamos varios indicadores, puede ser más claro organizarlos en una pequeña tabla.

Esto no cambia los resultados, pero mejora la lectura y acerca el análisis a una forma más comunicable.

In [38]:
indicadores_generales = pd.DataFrame({
    "indicador": [
        "Total facturado",
        "Total de propinas",
        "Promedio de cuenta",
        "Promedio de propina"
    ],
    "valor": [
        total_facturado,
        total_propinas,
        promedio_cuenta,
        promedio_propina
    ]
})

indicadores_generales

,indicador,valor
0,Total facturado,4827.770000
1,Total de propinas,731.580000
2,Promedio de cuenta,19.785943
3,Promedio de propina,2.998279


Ahora los indicadores aparecen organizados en una tabla.

Esta forma de presentación es más clara que una salida con varios valores sueltos, porque cada número queda asociado a su significado.

Sin embargo, los valores todavía aparecen con muchos decimales. Para comunicar mejor estos resultados, podemos redondearlos a dos decimales.

## Redondear resultados

Cuando los indicadores tienen muchos decimales, la lectura puede volverse innecesariamente pesada.

En este caso, redondear a dos decimales alcanza para comunicar los resultados de manera clara.

In [39]:
indicadores_generales["valor"] = indicadores_generales["valor"].round(2)

indicadores_generales

,indicador,valor
0,Total facturado,4827.77
1,Total de propinas,731.58
2,Promedio de cuenta,19.79
3,Promedio de propina,3.00


La tabla final presenta los indicadores generales de manera más clara.

Podemos resumir esta primera lectura así:

- En el dataset se registró un total facturado de **4827.77**.
- El total de propinas fue de **731.58**.
- El promedio de cuenta fue de **19.79**.
- El promedio de propina fue de **3.00**.

Estos indicadores permiten tener una primera idea general del dataset. Sin embargo, todavía no muestran diferencias internas.

Por ejemplo, no sabemos si el promedio de cuenta cambia entre almuerzo y cena, si las propinas varían según el día o si las mesas más grandes generan cuentas más altas. Para responder ese tipo de preguntas vamos a necesitar comparar grupos.

## Comparar una pregunta en dos subconjuntos

Hasta ahora calculamos indicadores para todo el dataset.

Pero una pregunta de análisis muchas veces necesita comparar partes del conjunto de datos. Por ejemplo:

**¿El promedio de cuenta es igual en el almuerzo y en la cena?**

Todavía no necesitamos una herramienta nueva para responder esto. Podemos usar filtros, que ya conocemos, para separar las cuentas de almuerzo y cena, y luego calcular el promedio de `total_bill` en cada caso.

In [40]:
promedio_almuerzo = df[df["time"] == "Lunch"]["total_bill"].mean()
promedio_cena = df[df["time"] == "Dinner"]["total_bill"].mean()

promedio_almuerzo, promedio_cena

(np.float64(17.168676470588235), np.float64(20.79715909090909))

El promedio de cuenta durante el almuerzo es aproximadamente **17.17**.

El promedio de cuenta durante la cena es aproximadamente **20.80**.

En este primer cálculo, las cuentas de la cena tienen un promedio más alto que las cuentas del almuerzo.

De todos modos, la salida vuelve a aparecer como dos valores sueltos. Para comparar mejor, conviene organizar el resultado en una pequeña tabla.

## Organizar la comparación

Cuando comparamos dos valores, una tabla simple puede ayudar a leer mejor el resultado.

Vamos a construir un pequeño `DataFrame` con el promedio de cuenta para cada horario.

In [41]:
comparacion_horario = pd.DataFrame({
    "horario": ["Lunch", "Dinner"],
    "promedio_total_cuenta": [promedio_almuerzo, promedio_cena]
})

comparacion_horario["promedio_total_cuenta"] = comparacion_horario["promedio_total_cuenta"].round(2)

comparacion_horario

,horario,promedio_total_cuenta
0,Lunch,17.17
1,Dinner,20.80


La tabla permite comparar con más claridad el promedio de cuenta según el horario.

En este dataset, las cuentas registradas durante la cena (`Dinner`) tienen un promedio de **20.80**, mientras que las cuentas registradas durante el almuerzo (`Lunch`) tienen un promedio de **17.17**.

Esto sugiere que, en estos datos, el consumo promedio es mayor durante la cena.

Esta comparación se resolvió con filtros, una herramienta que ya conocemos. Más adelante vamos a estudiar formas más directas de resumir datos por grupos, especialmente con `groupby()`.

## Una respuesta puede abrir nuevas preguntas

Un análisis exploratorio no avanza en línea recta. Muchas veces, una respuesta genera nuevas preguntas.

Por ejemplo, acabamos de observar que el promedio de cuenta es mayor durante la cena que durante el almuerzo. A partir de eso podríamos preguntarnos:

- ¿Hay muchas más cuentas de cena que de almuerzo?
- ¿La propina promedio también es mayor durante la cena?
- ¿El tamaño de las mesas cambia según el horario?

En este capítulo no vamos a responder todas esas preguntas. Pero sí podemos hacer una comparación adicional sencilla: revisar si la propina promedio también cambia entre almuerzo y cena.

In [42]:
promedio_propina_almuerzo = df[df["time"] == "Lunch"]["tip"].mean()
promedio_propina_cena = df[df["time"] == "Dinner"]["tip"].mean()

comparacion_propinas = pd.DataFrame({
    "horario": ["Lunch", "Dinner"],
    "promedio_propina": [promedio_propina_almuerzo, promedio_propina_cena]
})

comparacion_propinas["promedio_propina"] = comparacion_propinas["promedio_propina"].round(2)

comparacion_propinas

,horario,promedio_propina
0,Lunch,2.73
1,Dinner,3.10


La propina promedio también es mayor durante la cena.

En el almuerzo (`Lunch`), la propina promedio es de **2.73**. En la cena (`Dinner`), la propina promedio es de **3.10**.

Este resultado acompaña lo que habíamos visto con el total de la cuenta: en este dataset, tanto el promedio de cuenta como el promedio de propina son más altos durante la cena.

Este tipo de comparación empieza a mostrar una necesidad nueva: resumir información por grupos de manera más directa y ordenada.

## Cierre del capítulo

En este capítulo comenzamos a trabajar con una idea central del análisis de datos: convertir un dataset en preguntas.

Usamos un dataset de práctica, revisamos su tamaño y sus variables principales, contamos valores en columnas categóricas y calculamos indicadores simples sobre columnas numéricas.

También vimos que una respuesta inicial puede abrir nuevas preguntas. Al observar que la cena tenía un promedio de cuenta más alto que el almuerzo, surgió naturalmente la necesidad de comparar otros indicadores, como la propina promedio.

Hasta ahora hicimos comparaciones sencillas usando filtros y cálculos directos. Este enfoque es útil para empezar, pero puede volverse incómodo cuando queremos comparar muchas categorías o calcular varios indicadores por grupo.

En los próximos capítulos vamos a profundizar en formas más ordenadas de resumir información, construir tablas de análisis y comparar grupos dentro de un dataset.